lib/c_backend/transform.ml

```diff
@@ -685,6 +685,103 @@ and prepend_expr env ~prepend_stmts ~append_stmts (expr: expr_result) =
+  let open Expression in
+
+  let do_integer_folding int_op left_int right_int =
+    let new_int =
+      match int_op with
+      (* TODO: i64 support *)
+      | BinaryOp.Minus -> Int32.(-) left_int right_int
+      | BinaryOp.Plus -> Int32.(+) left_int right_int
+      | BinaryOp.Div -> Int32.(/) left_int right_int
+      | BinaryOp.Mult -> Int32.( * ) left_int right_int
+      | _ -> failwith "unsupported binary op between integer"
+    in
+    Constant(Literal.Integer(new_int))
+  in
+
+  let do_float_folding float_op left_str right_str =
+    let new_float =
+      let left_f = Float.of_string left_str in
+      let right_f = Float.of_string right_str in
+      match float_op with
+      | BinaryOp.Minus -> Float.(-) left_f right_f
+      | BinaryOp.Plus -> Float.(+) left_f right_f
+      | BinaryOp.Div -> Float.(/) left_f right_f
+      | BinaryOp.Mult -> Float.( * ) left_f right_f
+      | _ -> failwith "unsupported binary op between float"
+    in
+    Constant(Literal.Float(Float.to_string new_float, None))
+  in
+
+  let is_same_expr expr_left expr_right = (expr_left.ty_var = expr_right.ty_var) in
+
+  match (left.spec, right.spec) with
+  | (Constant(Integer left_int), Constant(Integer right_int)) -> (
+    match op with
+    | BinaryOp.Plus
+    | BinaryOp.Minus
+    | BinaryOp.Div
+    | BinaryOp.Mult -> (
+      let new_spec = do_integer_folding op left_int right_int in
+      { left with spec = new_spec }
+    )
+    | _ -> original_expr
+  )
+  | (Constant(Float(left_str, _)), Constant(Float(right_str, _))) -> (
+    match op with
+    | BinaryOp.Plus
+    | BinaryOp.Minus
+    | BinaryOp.Div
+    | BinaryOp.Mult -> (
+      let new_spec = do_float_folding op left_str right_str in
+      { left with spec = new_spec }
+    )
+    | _ -> original_expr
+  )
+  | (Constant(String(left_str, _, _)), Constant(String(right_str, _, _))) -> (
+    match op with 
+    | BinaryOp.Plus -> (
+      let concated = String.concat [left_str; right_str] in
+      let new_left_spec = Constant(Literal.String(concated, left.loc, None)) in
+      { left with spec = new_left_spec }
+    )
+    | _ -> original_expr
+  )
+  | (Binary(_, _, _), _) -> (
+    let left' = transform_binary_expression left in
+    if not (is_same_expr left' left) then
+      (* if left being transformed, try rec folding left' with right *)
+      do_literal_folding op left' right { original_expr with spec = Binary(op, left', right) }
+    else
+      original_expr
+  )
+  | (_, Binary(_, _, _)) -> (
+    let right' = transform_binary_expression right in
+    if not (is_same_expr right' right) then
+      (* if right being transformed, try rec folding left with right' *)
+      do_literal_folding op left right' { original_expr with spec = Binary(op, left, right') }
+    else
+      original_expr
+  )
+  | _ ->
+    (* pattern matching failed. propagate original_expr to parent *)
+    original_expr
+
+and transform_binary_expression expr: Expression.t =
```

Comment:
The name of this function misleads me. Because in this file `transform` represents the procedure of turning `Expression.t` into `C_op.Exp.t`, but this function doesn't.
I would suggest using a better name.